packages/scripts/src/upgrade-tests/register-new-contract/1_Propose.ts

```diff
@@ -0,0 +1,198 @@
+  Finder,
+  Governor,
+  GovernorHub,
+  GovernorRootTunnel,
+  Proposer,
+  Registry,
+} from "@uma/contracts-node/typechain/core/ethers";
+import { BaseContract, PopulatedTransaction } from "ethers";
+import { getContractInstance, getContractInstanceByUrl } from "../../utils/contracts";
+import { fundArbitrumParentMessengerForRelays, relayGovernanceMessages } from "../../utils/relay";
+
+// PARAMETERS
+const proposerWallet = "0x2bAaA41d155ad8a4126184950B31F50A1513cE25";
+const newContractName = interfaceName.OptimisticAsserter;
+
+// Env vars
+const NODE_URL_ENV = "NODE_URL_";
+
+async function main() {
+  const proposerSigner = await hre.ethers.getSigner(proposerWallet);
+
+  const newContractAddressMainnet = await getAddress(newContractName, 1);
+
+  const finder = await getContractInstance<Finder>("Finder");
+  const governor = await getContractInstance<Governor>("Governor");
+  const registry = await getContractInstance<Registry>("Registry");
+  const proposer = await getContractInstance<Proposer>("Proposer");
+  const arbitrumParentMessenger = await getContractInstance<ArbitrumParentMessenger>("Arbitrum_ParentMessenger");
+
+  const governorRootTunnel = await getContractInstance<GovernorRootTunnel>("GovernorRootTunnel"); // for polygon
+  const governorHub = await getContractInstance<GovernorHub>("GovernorHub"); // rest of l2
+
+  const l2Networks = { BOBA: 288, MATIC: 137, OPTIMISM: 10, ARBITRUM: 42161 };
+
+  const adminProposalTransactions: {
+    to: string;
+    value: BigNumberish;
+    data: BytesLike;
+  }[] = [];
+
+  if (!newContractAddressMainnet) throw new Error(`No ${newContractName} address found in mainnet deployment`);
+
+  for (const networkName in l2Networks) {
+    const governanceMessages: { targetAddress: string; tx: PopulatedTransaction }[] = [];
+    const l2ChainId = l2Networks[networkName as keyof typeof l2Networks];
+    const l2NodeUrl = process.env[String(NODE_URL_ENV + l2ChainId)];
+    const l2NewContractAddress = await getAddress(newContractName, l2ChainId);
+
+    if (!l2NodeUrl || !l2NewContractAddress) throw new Error(`Missing ${networkName} network config`);
+
+    const isPolygon = l2ChainId === 137;
+    const isArbitrum = l2ChainId === 42161;
+
+    const l2Registry = await getContractInstanceByUrl<Registry>("Registry", l2NodeUrl);
+
+    // The l2Governor in polygon is the GovernorChildTunnel and in the rest of the l2's is the GovernorHub
+    const l2Governor = await getContractInstanceByUrl<BaseContract>(
+      isPolygon ? "GovernorChildTunnel" : "GovernorSpoke",
+      l2NodeUrl
+    );
+    const l2Finder = await getContractInstanceByUrl<Finder>("Finder", l2NodeUrl);
+
+    if (await l2Registry.isContractRegistered(l2NewContractAddress)) continue;
+
+    console.log(`Registering ${l2NewContractAddress} on ${networkName}`);
+
+    // Fund Arbitrum if needed for next 4 transactions
+    if (isArbitrum) await fundArbitrumParentMessengerForRelays(arbitrumParentMessenger, proposerSigner, 5);
```

Comment:
kinda sucks that we need to do this for these two chains but it's not too bad. you could pre-fund the arbitrum setup  to avoid needing to do this but I think it's better to show the full flow within the script.

Reply:
Yes, it's required for this scripts to work, even when running them in mainnet, we still need to fund the arbitrumParentMessenger (it doesn't need to be during proposals though, but before the execution)